universal_transfer_operator/src/universal_transfer_operator.py

```diff
@@ -0,0 +1,78 @@
+from transfers.datasets.base import UniversalDataset as Dataset
+from transfers.integrations import get_transfer_integration
+from transfers.utils import check_if_connection_exists
+
+from astro.constants import LoadExistStrategy
+
+
+class UniversalTransferOperator(BaseOperator):
+    """
+    Transfers all the data that could be read from the source Dataset into the destination Dataset. From a DAG author
+    standpoint, all transfers would be performed through the invocation of only the Universal Transfer Operator.
+
+    :param source_dataset: Source dataset to be transferred.
+    :param destination_dataset: Destination dataset to be transferred to.
+    :param transfer_params: kwargs to be used by method involved in transfer flow.
+    :param transfer_mode: Use transfer_mode TransferMode; native, non-native or thirdparty.
+    :param if_exists: Overwrite file if exists. Default False.
+
+    :return: returns the destination dataset
+    """
+
+    def __init__(
+        self,
+        *,
+        source_dataset: Dataset,
+        destination_dataset: Dataset,
+        transfer_params: dict | None = {},
+        transfer_mode: TransferMode | None = TransferMode.NONNATIVE,
+        if_exists: LoadExistStrategy = "replace",
+        **kwargs,
+    ) -> None:
+        self.source_dataset = source_dataset
+        self.destination_dataset = destination_dataset
+        self.transfer_mode = transfer_mode or TransferMode.NONNATIVE
+        # TODO: revisit names of transfer_mode
+        self.transfer_params = transfer_params or {}
+        self.if_exists = if_exists
+        super().__init__(**kwargs)
+
+    def execute(self, context: Context) -> Any:
+        if self.source_dataset.conn_id:
+            if not check_if_connection_exists(self.source_dataset.conn_id):
+                raise ValueError("source_dataset connection does not exist.")
+
+        if self.destination_dataset.conn_id:
+            if not check_if_connection_exists(self.source_dataset.conn_id):
+                raise ValueError("destination_dataset connection does not exist.")
+
+        if self.transfer_mode is TransferMode.THIRDPARTY:
+            transfer_integration = get_transfer_integration(self.transfer_params)
+            return transfer_integration.transfer_job(self.source_dataset, self.destination_dataset)
+
+        source_dataprovider = create_dataprovider(
+            dataset=self.source_dataset,
+            transfer_params=self.transfer_params,
+            transfer_mode=self.transfer_mode,
+            if_exists=self.if_exists,
+        )
+        source_data = source_dataprovider.read()
+
+        destination_dataprovider = create_dataprovider(
+            dataset=self.destination_dataset,
+            transfer_params=self.transfer_params,
+            transfer_mode=self.transfer_mode,
+            if_exists=self.if_exists,
+        )
+
+        destination_data = destination_dataprovider.write(source_data)
+        cleanup(source_data)
```

Comment:
It would be great if we could do the actual transfer in a context manager (with `try`, `finally`). So that we `cleanup` also when an error occurs. Something like:
```python
with source_dataprovider.read() as source_data:
    destination_dataprovider.write(source_data)
```

Reply:
+1

Reply:
+1

Reply:
Added the context manager here: https://github.com/astronomer/astro-sdk/pull/1492/commits/831d627d1a86acab300968fc56d11bc7f016b4b0 Will add the logic for actual transfer in further commits.

Reply:
Using this instead: https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager

```
from contextlib import contextmanager

@contextmanager
def managed_resource(*args, **kwds):
    # Code to acquire resource, e.g.:
    resource = acquire_resource(*args, **kwds)
    try:
        yield resource
    finally:
        # Code to release resource, e.g.:
        release_resource(resource)
```

Reply:
https://github.com/astronomer/astro-sdk/pull/1492/commits/ac3af739b548c86636c4519e0c1c7bcadf321ead

Reply:
https://github.com/astronomer/astro-sdk/pull/1492/commits/22b0aeeb515b6e9eb72dd68f62a64cb807a5732a